File: SecondPhaseGate(3in1)Java/AtmMachineApp.java
public class AtmMachineApp {

    private final String firstName;
    private final String lastName;
    private int pin;
    private double balance;
    private boolean isClosed;

    public AtmMachineApp(String firstName, String lastName, int pin) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.pin = pin;
        this.balance = 0;
        this.isClosed = false;
    }

    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public int getPin() {
        return pin;
    }

    public double getBalance() {
        return balance;
    }

    public boolean isClosed() {
        return isClosed;
    }

    public void closeAccount() {
        isClosed = true;
    }

    public void depositMoney(int pin, double amount) {
        if(this.pin == pin && !isClosed) {
            balance += amount;
        }else if(this.pin != pin) {
            throw new IllegalArgumentException("Invalid pin");
        } else {
            throw new IllegalArgumentException("Account closed");
        }
    }


    public void withdrawMoney(int pin, double amount) {
        if(this.pin == pin && !isClosed && this.balance >= amount) {
            balance -= amount;
        } else if(this.pin != pin) {
            throw new IllegalArgumentException("Invalid pin");
        }else if(isClosed) {
            throw new IllegalArgumentException("Account closed");
        } else if(this.balance >= amount) {
            throw new IllegalArgumentException("Insufficient balance");
        }
    }

    public void transferMoney(AtmMachineApp account2, double amount, int pin) {
      if(this.pin == pin && !isClosed && this.balance >= amount) {
          this.withdrawMoney(pin, amount);
          account2.depositMoney(pin,amount);
      }else if(this.pin != pin) {
          throw new IllegalArgumentException("Invalid pin");
      }else if(isClosed) {
          throw new IllegalArgumentException("Account closed");
      }
    }

    public void changePin(int oldPin, int newPin) {
        if (this.pin == oldPin) {
            this.pin = newPin;
        } else {
            throw new IllegalArgumentException("Invalid pin");
        }
    }

}

File: SecondPhaseGate(3in1)Python/atm_machine_app.py

class atm_machine_app:
    def __init__(self, first_name, last_name, pin):
        self._first_name = first_name
        self._last_name = last_name
        self._pin = pin
        self._balance = 0
        self._is_closed = False

    @property
    def first_name(self):
        return self._first_name

    @property
    def last_name(self):
        return self._last_name

    @property
    def pin(self):
        return self._pin

    @property
    def balance(self):
        return self._balance

    @property
    def is_closed(self):
        return self._is_closed

    def close_account(self):
        self._is_closed = True

    def deposit_money(self, pin, amount):
        if amount <= 0:
            raise ValueError("Amount must be greater than 0.")
        if self._pin == pin and not self._is_closed:
            self._balance += amount
        elif self._pin != pin:
            raise ValueError("Invalid PIN")
        else:
            raise ValueError("Account closed")

    def withdraw_money(self, pin, amount):
        if self._pin == pin and not self._is_closed:
            self._balance -= amount
        elif self._pin != pin:
            raise ValueError("Invalid PIN")

    def transfer_money(self, account2, amount, pin):
        if self._pin == pin and not self._is_closed and self._balance >= amount:
            self.withdraw_money(pin, amount)
            account2.deposit_money(account2.pin, amount)
        elif self._pin != pin:
            raise ValueError("Invalid PIN")
        elif self._is_closed:
            raise ValueError("Account closed")
        elif self._balance < amount:
            raise ValueError("Insufficient balance")

    def change_pin(self, oldPin, newPin):
        if self._pin == oldPin and not self._is_closed:
            self._pin = newPin
        else:
            raise ValueError("Invalid PIN")





